### PR TITLE
In progress jobs and general refactoring

### PR DESCRIPTION
**Pull request by rizo**
This PR focuses on improving the structure of the codebase and introduces the ability to track benchmarking job status with persistence in the database.

More precisely:

- The database schema is updated to track the job status based on the presence of job identifiers for build and run stages provided by current.
  - I might revisit this encoding in a follow up PR. Having a separate field for the status could make things easier for the frontend.
- Refactor the application entry-point (`main.ml`) to simplify config and source initialisation.
- Improve config management by introducing a centralised module for config options.
- Add migration scripts from the existing schema to the new schema with support for in-progress jobs.
  - My suggestion would be to deploy this PR as is to our testing environment, before applying the migrations.
- Introduce a shared "commit context" abstraction that precisely describes the commit the benchmarks will be run for.
- Simplify the storage module and add additional SQL statements to track job status updates.
  - I have a follow-up that reworks the db code by using Caqti.

I also suggest that we create and merge this PR into a `development` branch instead directly merging this into `main`. This would help us make and test changes in the development environment.

### CODE REVIEW

File: pipeline/lib/pipeline.ml [hunk 1]
```diff
@@ -74,79 +76,109 @@ let github_status_of_state url = function
...
-               |> Yojson.Safe.Util.to_string_option
-             in
-             Yojson.Safe.Util.(member "results" output_json)
-             |> Yojson.Safe.Util.to_list
-             |> List.map
-                  (Benchmark.make ~duration ~run_at ~repo_id ~benchmark_name
-                     ~commit ?pull_number ?build_job_id ?run_job_id ?branch)
-             |> List.iter (Models.Benchmark.Db.insert db));
-      db#finish
+    let* run_job_id = Current_util.get_job_id current_output
```
Comment (gs0510) [inline pipeline/lib/pipeline.ml:126]: will `run_job_id` only be determined once docker has run with the make bench target? Would we still have a `pending stage`? I think the only thing left to do here is to put the pread_logs in the database.

**Review by gs0510 — COMMENTED** [review]
This is great @rizo! Left some comments about things I'm not clear on. :)

File: pipeline/bin/main.ml [hunk 2]
```diff
@@ -15,66 +30,64 @@ module Github = struct
...
        enabled, see `lscpu`)"
     in
     Arg.(value & opt (some int) None & info [ "docker-numa-node" ] ~doc)
 
-  let shm_size =
+  let docker_shm_size =
     let doc = "Size of tmpfs volume to be mounted in /dev/shm (in GB)." in
     Arg.(value & opt int 4 & info [ "docker-shm-size" ] ~doc)
 
+  let slack_path =
```
Comment (gs0510) [inline pipeline/bin/main.ml:60]: Should we remove this? We don't use this anymore and I don't think we have any plans of posting anything to slack in the future.

File: pipeline/lib/engine.ml [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    @@ run "opam remote add origin https://opam.ocaml.org"
+    @@ run "%s" opam_dependencies
+    @@ add ~src:[ "--chown=opam ." ] ~dst:"." ()
+    @@ run "eval $(opam env)"
+
+  let build ~pool (_commit_context : Commit_context.t) commit =
+    let dockerfile =
+      let+ base = Docker.pull ~schedule:weekly "ocaml/opam" in
+      (* TODO *)
+      `Contents (dockerfile ~base ~repository:"x")
```
Comment (gs0510) [inline pipeline/lib/engine.ml:95]: I think this can just be `_commit_context.repo_id_string`?

**Review by gs0510 — COMMENTED** [review]
This is amazing @rizo! I have some comments, but I think I need one more pass to understand things.

One question I also have that I think I asked before was how would the frontend poll for the changes in the state? Would it mean we need to do more graphql queries in the frontend?

**Comment by rizo** [comment]
Thanks for the feedback, @gs0510! Indeed, I noticed a few incomplete things when I started testing this. I'm already working on improvements.

I'll let you know when I address all of your comments.

**Comment by rizo** [comment]
Also it's very nice having current-bench running for current-bench! Thanks for setting it up! :)

**Comment by rizo** [comment]
@gs0510 this should be now ready for a (hopefully!) final review and testing.

If you want to test it locally:

1. Manually create the `benchmarks_new` table found in `./db/migrations/20210411135924_populate_benchmarks_new_table.up.sql`.

2. Start the pipeline:

```
$ dune exec --root=. -- bin/main.exe local --verbosity=debug --conn-info="postgresql://docker:docker@localhost:5432/docker" ../local-test-repo/
```

3. Make and commit any changes in the testing sub-repository `./local-test-repo` (you might need to initialize it).


I am not currently testing this with docker-compose because we still need to update the UI to use the new schema. I'll start working on that in my next PR.

File: pipeline/db/migrations/20210411132055_create_benchmarks_new_table.up.sql [hunk 4]
```diff
@@ -0,0 +1,16 @@
...
+
+CREATE TABLE benchmarks_new (
+  started_at TIMESTAMP WITHOUT TIME ZONE DEFAULT CURRENT_TIMESTAMP,
+  repo_id CHARACTER VARYING(256) NOT NULL,
+  commit CHARACTER VARYING(50) NOT NULL,
+  branch CHARACTER VARYING(256) DEFAULT NULL,
+  pull_number INTEGER DEFAULT NULL,
+  build_job_id CHARACTER VARYING(256) NOT NULL,
+  run_job_id CHARACTER VARYING(256) DEFAULT NULL,
+  status json DEFAULT NULL,
```
Comment (gs0510) [inline pipeline/db/migrations/20210411132055_create_benchmarks_new_table.up.sql:11]: Why did you change the status to be json?

**Review by gs0510 — APPROVED** [review]
Thanks for the instructions in the last comment! LGTM! :) I just have one question :)

File: pipeline/db/migrations/20210411132055_create_benchmarks_new_table.up.sql (hunk 4 above)
Comment (rizo) [reply pipeline/db/migrations/20210411132055_create_benchmarks_new_table.up.sql:11]: This way we can store errors in a more uniform way and present something useful in the UI. See: https://github.com/ocurrent/current-bench/pull/120/files#diff-9769a3f5504af3cd735e08f078b94f827bdd51eacac2e9d80ac9761bc485171fR249

**Comment by rizo** [comment]
@gs0510 do you mind setting up a `development` branch to merge this PR (I don't think I have the right permissions)?

**Comment by gs0510** [comment]
@rizo pushed a `development` branch and merged! :)

**Comment by gs0510** [comment]
@rizo  Should I deploy this on the dev server? Or do you want to do it? :)

File: pipeline/db/migrations/20210411135924_populate_benchmarks_new_table.up.sql [hunk 5]
```diff
@@ -0,0 +1,34 @@
+-- WIP
```
Comment (rizo) [inline pipeline/db/migrations/20210411135924_populate_benchmarks_new_table.up.sql:1]: This is a WIP implementation of the migration to move the existing `benchmarks` entries into `benchmarks_new` schema.

I'm still testing this and will finish it in a follow-up PR.
